### PR TITLE
Add battery status from ESCs, refactor battery source param

### PR DESCRIPTION
**Pull request by julianoes**
This PR includes two things:

1. Addition of `esc_battery` module to publish the `battery_status` from `esc_status`.
2. Refactoring of how `BAT%d_SOURCE` param is used. The param is now checked inside of the battery library. If the source set matches the data fed to the library, then it is published, otherwise it isn't.

### CODE REVIEW

File: msg/battery_status.msg [hunk 1]
```diff
@@ -10,8 +10,12 @@ float32 scale		# Power scaling factor, >= 1, or -1 if unknown
 float32 temperature         # temperature of the battery. NaN if unknown
 int32 cell_count		# Number of cells
 bool connected			# Whether or not a battery is connected, based on a voltage threshold
-bool system_source		# Whether or not a this battery is the active power source for VDD_5V_IN
-uint8 priority        	# Zerro based priority is the connection on the Power Controller V1..Vn AKA BrickN-1
+
+uint8 BATTERY_SOURCE_POWER_MODULE = 0
+uint8 BATTERY_SOURCE_EXTERNAL = 0
+uint8 BATTERY_SOURCE_ESCS = 0
```
Comment (dakejahl) [inline msg/battery_status.msg]: 0, 1, 2 ?

File: src/drivers/power_monitor/ina226/ina226.cpp [hunk 2]
```diff
@@ -88,10 +88,9 @@ INA226::INA226(I2CSPIBusOption bus_option, const int bus, int bus_frequency, int
 		0.0,
 		0.0,
 		false,
-		false, // TODO: selected source?
+		battery_status_s::BATTERY_SOURCE_POWER_MODULE,
 		0,
-		0.0,
-		true
+		0.0
 	);
```
Comment (dakejahl) [inline src/drivers/power_monitor/ina226/ina226.cpp]: yuck, this function has 8 arguments LOL

File: msg/battery_status.msg (hunk 1 above)
Comment (dakejahl) [reply msg/battery_status.msg]: I don't see where the BAT%d_SOURCE parameter definitions are located? Could you point me to them and make sure the documentation is updated?

Comment (julianoes) [reply msg/battery_status.msg]: Oops!

Comment (julianoes) [reply msg/battery_status.msg]: Fixed in a0c82b1.

File: src/drivers/power_monitor/ina226/ina226.cpp (hunk 2 above)
Comment (julianoes) [reply src/drivers/power_monitor/ina226/ina226.cpp]: :confused:

File: src/lib/battery/module.yaml [hunk 3]
```diff
@@ -129,12 +129,13 @@ parameters:
                     This parameter controls the source of battery data. The value 'Power Module'
                     means that measurements are expected to come from a power module. If the value is set to
                     'External' then the system expects to receive mavlink battery status messages.
+                    If the value is set to 'ESCs', the battery information are taken from the esc_status message.
```
Comment (bkueng) [inline src/lib/battery/module.yaml]: Might be worth mentioning that this includes both voltage and current (and therefore the ESC must provide both).

File: src/modules/battery_status/analog_battery.cpp [hunk 4]
```diff
@@ -127,6 +127,4 @@ AnalogBattery::updateParams()
 			param_set_no_notification(_analog_param_handles.a_per_v_old, &_analog_params.a_per_v_old);
 		}
 	}
-
-	Battery::updateParams();
```
Comment (bkueng) [inline src/modules/battery_status/analog_battery.cpp]: Why is this removed?

File: src/lib/battery/battery.cpp [hunk 5]
```diff
@@ -339,7 +341,5 @@ void Battery::updateParams()
 	param_get(_param_handles.crit_thr, &_params.crit_thr);
 	param_get(_param_handles.emergen_thr, &_params.emergen_thr);
 
-	ModuleParams::updateParams();
```
Comment (bkueng) [inline src/lib/battery/battery.cpp]: Also here: why is this removed?

File: src/drivers/uavcan/sensors/battery.cpp [hunk 6]
```diff
@@ -95,7 +95,8 @@ UavcanBatteryBridge::battery_sub_cb(const uavcan::ReceivedDataStructure<uavcan::
 	// battery.average_time_to_empty = msg.;
 	battery.serial_number = msg.model_instance_id;
 	battery.connected = true;
-	battery.system_source = msg.status_flags & uavcan::equipment::power::BatteryInfo::STATUS_FLAG_IN_USE;
+	battery.source =
```
Comment (bkueng) [inline src/drivers/uavcan/sensors/battery.cpp]: Why 2 lines?

File: src/modules/battery_status/analog_battery.cpp [hunk 7]
```diff
@@ -40,8 +40,8 @@ AnalogBattery::AnalogBattery(int index, ModuleParams *parent) :
 }
 
 void
-AnalogBattery::updateBatteryStatusADC(hrt_abstime timestamp, float voltage_raw, float current_raw,
-				      bool selected_source, int priority, float throttle_normalized)
+AnalogBattery::updateBatteryStatusADC(hrt_abstime timestamp, int32_t voltage_raw, int32_t current_raw,
```
Comment (bkueng) [inline src/modules/battery_status/analog_battery.cpp]: Did you really intend to change to int32_t?

File: src/modules/esc_battery/EscBattery.cpp [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+EscBattery::EscBattery() :
+	ModuleParams(nullptr),
+	WorkItem(MODULE_NAME, px4::wq_configurations::lp_default),
+	_battery(1, this)
+{
+	parameters_updated();
+}
+
+EscBattery::~EscBattery()
```
Comment (bkueng) [inline src/modules/esc_battery/EscBattery.cpp]: default destructor

File: src/modules/esc_battery/EscBattery.cpp [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+#include <drivers/drv_hrt.h>
+
+using namespace time_literals;
+
+EscBattery::EscBattery() :
+	ModuleParams(nullptr),
+	WorkItem(MODULE_NAME, px4::wq_configurations::lp_default),
+	_battery(1, this)
+{
+	parameters_updated();
```
Comment (bkueng) [inline src/modules/esc_battery/EscBattery.cpp]: You can remove this

File: src/modules/esc_battery/EscBattery.cpp [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+		parameters_updated();
+	}
+
+	esc_status_s esc_status;
+
+	if (_esc_status_sub.update(&esc_status)) {
+
+		if (hrt_elapsed_time(&esc_status.timestamp) > 500_ms ||
+		    esc_status.esc_count == 0 ||
+		    esc_status.esc_count > 8) {
```
Comment (bkueng) [inline src/modules/esc_battery/EscBattery.cpp]: Why hardcode to 8?

File: src/modules/esc_battery/EscBattery.cpp [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+		// Clear update
+		parameter_update_s param_update;
+		_parameter_update_sub.copy(&param_update);
+
+		parameters_updated();
+	}
+
+	esc_status_s esc_status;
+
+	if (_esc_status_sub.update(&esc_status)) {
```
Comment (bkueng) [inline src/modules/esc_battery/EscBattery.cpp]:
```suggestion
	if (_esc_status_sub.copy(&esc_status)) {
```

File: src/modules/esc_battery/EscBattery.cpp [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+		if (hrt_elapsed_time(&esc_status.timestamp) > 500_ms ||
```
Comment (bkueng) [inline src/modules/esc_battery/EscBattery.cpp]: You can add this but don't have to since you already know the topic got updated at this point.

File: src/modules/battery_status/analog_battery.cpp (hunk 4 above)
Comment (julianoes) [reply src/modules/battery_status/analog_battery.cpp]: @bkueng can you check if my commit message here makes sense?
https://github.com/PX4/Firmware/pull/14490/commits/d819c52ff12e599a259663f05e46269973ef1c41

Comment (bkueng) [reply src/modules/battery_status/analog_battery.cpp]: But this is not how inheritance works, or rather the pattern that is used here. What we have here is a subclass overriding a parent's implementation, because it has some customizations, and then calls the parent's method.

Comment (julianoes) [reply src/modules/battery_status/analog_battery.cpp]: Ok, in that case I'll revert this.

File: src/drivers/uavcan/sensors/battery.cpp (hunk 6 above)
Comment (julianoes) [reply src/drivers/uavcan/sensors/battery.cpp]: Thx fixed.

File: src/modules/battery_status/analog_battery.cpp (hunk 7 above)
Comment (julianoes) [reply src/modules/battery_status/analog_battery.cpp]: Huh, I did not fix that merge conflict correctly, thanks!

File: src/modules/esc_battery/EscBattery.cpp (hunk 11 above)
Comment (julianoes) [reply src/modules/esc_battery/EscBattery.cpp]: Right :man_facepalming:

File: src/modules/esc_battery/EscBattery.cpp (hunk 12 above)
Comment (julianoes) [reply src/modules/esc_battery/EscBattery.cpp]: True. Removing it.

File: src/modules/esc_battery/EscBattery.cpp (hunk 10 above)
Comment (julianoes) [reply src/modules/esc_battery/EscBattery.cpp]: I'm changing it to the define from the msg file.

**Comment by julianoes** [comment]
@bkueng everything should be addressed.
@dagar or @AlexKlimaj I'd appreciate your review.

File: src/drivers/uavcan/sensors/battery.cpp [hunk 13]
```diff
@@ -95,7 +95,7 @@ UavcanBatteryBridge::battery_sub_cb(const uavcan::ReceivedDataStructure<uavcan::
 	// battery.average_time_to_empty = msg.;
 	battery.serial_number = msg.model_instance_id;
 	battery.connected = true;
-	battery.system_source = msg.status_flags & uavcan::equipment::power::BatteryInfo::STATUS_FLAG_IN_USE;
+	battery.source = battery_status_s::BATTERY_SOURCE_POWER_MODULE;
```
Comment (AlexKlimaj) [inline src/drivers/uavcan/sensors/battery.cpp]: This is not true since a UAVCAN battery is not the power module. In the current implementation, this tells you if the battery is charging or discharging.

Comment (julianoes) [reply src/drivers/uavcan/sensors/battery.cpp:98]: Aha! What should I change it to?

Comment (AlexKlimaj) [reply src/drivers/uavcan/sensors/battery.cpp:98]: `battery.source = msg.status_flags & uavcan::equipment::power::BatteryInfo::STATUS_FLAG_IN_USE;`

Comment (julianoes) [reply src/drivers/uavcan/sensors/battery.cpp:98]: So I've removed the field `system_source` because it was not used anywhere from what I could tell and the description was not clear to me. This is now just a `source` field indicating what it is.

Comment (AlexKlimaj) [reply src/drivers/uavcan/sensors/battery.cpp:98]: I can only see the system_source being useful if a vehicle batteries that can be charged while attached.

Comment (julianoes) [reply src/drivers/uavcan/sensors/battery.cpp:98]: I would suggest to add it again if we have the implementation using it and the whole picture around it.

**Comment by julianoes** [comment]
@dakejahl I have fixed the enum.

**Comment by julianoes** [comment]
@bkueng are you ok if we merge this?
